src/services/impl/WindowsFileSystem.cpp

```diff
@@ -0,0 +1,65 @@
+#include "WindowsFileSystem.hh"
+
+#include "RA_Log.h"
+#include "RA_StringUtils.h"
+
+#include "services\impl\FileTextReader.hh"
+#include "services\impl\FileTextWriter.hh"
+
+#include <fstream>
+
+#define WIN32_LEAN_AND_MEAN
+struct IUnknown;
+#include <Windows.h>
+#include <Shlwapi.h>
+#undef CreateDirectory
+
+namespace ra {
+namespace services {
+namespace impl {
+
+WindowsFileSystem::WindowsFileSystem()
+{
+    // determine the home directory from the executable's path
+    wchar_t sBuffer[MAX_PATH];
+    GetModuleFileNameW(0, sBuffer, MAX_PATH);
+    PathRemoveFileSpecW(sBuffer);
+    m_sBaseDirectory = sBuffer;
+    if (m_sBaseDirectory.back() != '\\')
+        m_sBaseDirectory.push_back('\\');
+
+    RA_LOG("BaseDirectory: %s", ra::Narrow(m_sBaseDirectory).c_str());
+}
+
+bool WindowsFileSystem::DirectoryExists(const std::wstring& sDirectory) const
+{
+    DWORD nAttrib = GetFileAttributesW(sDirectory.c_str());
+    return (nAttrib != INVALID_FILE_ATTRIBUTES);
+}
+
+bool WindowsFileSystem::CreateDirectory(const std::wstring& sDirectory) const
+{
+    return static_cast<bool>(CreateDirectoryW(sDirectory.c_str(), nullptr));
+}
+
+std::unique_ptr<TextReader> WindowsFileSystem::OpenTextFile(const std::wstring& sPath) const
+{
+    auto pReader = std::make_unique<FileTextReader>(sPath);
+    if (!pReader->GetFStream().is_open())
+        return std::unique_ptr<TextReader>();
+
+    return std::unique_ptr<TextReader>(pReader.release());
```

Comment:
I'm not really sure if this OK or not. I'd recommend using `std::swap` with a new `std::unique_ptr` or moving `pReader `into the new one. Something like this...
```
auto pRet = std::move(pReader);
return pRet;
```
If you're sure it doesn't leak don't worry about it.

Reply:
this was required to change the pointer type. the first pointer is the subclass, and a pointer to the base class is returned. `release` detaches the pointer, so ownership can be reassigned to the other `unique_ptr`.